### PR TITLE
improve life with large tuples a little

### PR DESCRIPTION
**Pull request by carnaval**
I put the cutoff mentioned in #15702 in there.
tuple `_apply` inlining was unlimited.
Otherwise it's mostly trying to avoid llvm scalarizing load/store of large aggregates. Instead it uses the memcpy intrinsics so it is lowered to a function call when the size exceeds some thresold.
It's not very systematic, I'm just going through examples I see with catastrophic IR explosion while throwing random large tuples around.
Without this patch, just passing a large tuple through to the constructor of a type like

``` julia
immutable C
x :: NTuple{N,Int}
end
```

generates an amount IR linear wrt the size of the tuple.

I'm hoping this won't prevent optimizations on small sizes (we may want to add our own thresold for that if it happens). Anyway, I trust llvm more to break up a small memcpy than merge a lot of loads.

### CODE REVIEW

**Comment by carnaval** [comment]
just for fun, before :
```
$ time (./julia -e 'code_native(i->i, (NTuple{2000,Int},))' | wc -l)
4010

real	0m25.366s
user	0m25.289s
sys	0m0.285s
```
after
```
$ time (./julia -e 'code_native(i->i, (NTuple{2000,Int},))' | wc -l)
20

real	0m0.497s
user	0m0.497s
sys	0m0.210s
```

File: src/cgutils.cpp [hunk 1]
```diff
@@ -1442,6 +1459,14 @@ static Value *call_with_unsigned(Function *ufunc, Value *v)
 
 static void jl_add_linfo_root(jl_lambda_info_t *li, jl_value_t *val);
 
+static Value *as_value(Type *t, const jl_cgval_t &v)
+{
+    assert(!v.isboxed);
+    if (v.ispointer())
+        return tbaa_decorate(v.tbaa, build_load(builder.CreatePointerCast(v.V, t->getPointerTo()), v.typ));
```
Comment (vtjnash) [inline src/cgutils.cpp]: should use `data_pointer()` instead of `v.V` (in case `v` is a constant)

File: src/cgutils.cpp [hunk 2]
```diff
@@ -1442,6 +1459,14 @@ static Value *call_with_unsigned(Function *ufunc, Value *v)
 
 static void jl_add_linfo_root(jl_lambda_info_t *li, jl_value_t *val);
 
+static Value *as_value(Type *t, const jl_cgval_t &v)
+{
+    assert(!v.isboxed);
```
Comment (vtjnash) [inline src/cgutils.cpp]: this is faulty. the code-generator can't meet this condition.

File: src/cgutils.cpp [hunk 3]
```diff
@@ -1354,12 +1354,29 @@ static Value *emit_array_nd_index(const jl_cgval_t &ainfo, jl_value_t *ex, size_
...
     tbaa_decorate(tbaa, builder.CreateAlignedStore(v, builder.CreateBitCast(newv, PointerType::get(v->getType(),0)), sizeof(void*))); // min alignment in julia's gc is pointer-aligned
     return newv;
 }
+static Value *as_value(Type *t, const jl_cgval_t&);
+static Value *init_bits_cgval(Value *newv, const jl_cgval_t& v, MDNode *tbaa, Type *t)
+{
+    Value *jt = literal_pointer_val(v.typ);
+    if (v.ispointer()) {
+        init_tag(newv, jt);
+        builder.CreateMemCpy(newv, v.V, jl_datatype_size(v.typ), sizeof(void*));
```
Comment (vtjnash) [inline src/cgutils.cpp]: `data_pointer(v)`

File: src/cgutils.cpp [hunk 4]
```diff
@@ -1354,12 +1354,29 @@ static Value *emit_array_nd_index(const jl_cgval_t &ainfo, jl_value_t *ex, size_
...
+        return newv;
+    }
+    else {
+        return init_bits_value(newv, jt, as_value(t,v), tbaa);
```
Comment (vtjnash) [inline src/cgutils.cpp]: I think you might as well just use `v.V`, since you already checked that it is unboxed

File: src/intrinsics.cpp [hunk 5]
```diff
@@ -276,47 +276,70 @@ static Value *emit_unbox(Type *to, const jl_cgval_t &x, jl_value_t *jt)
...
-        load = build_load(p, jt);
+        alignment = julia_alignment(p, jt, 0);
     }
     else {
         // stack has default alignment
-        load = builder.CreateLoad(p);
+        alignment = 0;
+    }
+    if (dest) {
+        builder.CreateMemCpy(dest, p, jl_datatype_size(jt), alignment);
```
Comment (vtjnash) [inline src/intrinsics.cpp]: `, false, x.tbaa`

File: src/intrinsics.cpp [hunk 6]
```diff
@@ -257,8 +257,8 @@ static Constant *julia_const_to_llvm(jl_value_t *e, bool nested=false)
 
 static jl_cgval_t ghostValue(jl_value_t *ty);
 
-// emit code to unpack a raw value from a box into registers
-static Value *emit_unbox(Type *to, const jl_cgval_t &x, jl_value_t *jt)
+// emit code to unpack a raw value from a box into registers or a stack slot
+static Value *emit_unbox(Type *to, const jl_cgval_t &x, jl_value_t *jt, Value *dest)
```
Comment (vtjnash) [inline src/intrinsics.cpp]: `emit_unbox_or_don't` – well, that's a new one :)

File: src/cgutils.cpp (hunk 2 above)
Comment (vtjnash) [inline src/cgutils.cpp]: ah, nvm. i see you only use this after checking.

File: src/cgutils.cpp (hunk 1 above)
Comment (vtjnash) [inline src/cgutils.cpp]: this doesn't matter either then (since you already checked for `isconstant`)

**Pull request merged**